### DISCUSSION TITLE
Double Declaration error

### DISCUSSION THREAD

**Discussion by alexmkrolak** [visose/Robots#136] [Q&A] [answered]
I'm having an issue with a KR30 robot. When I upload the files I get 8 double declaration errors (error #: 2324) which all refer to the Global variables declared in the .DAT file. The SRC and DAT files are in their own folder in the R1/Program/ folder on the robot. Any tips?

[pp2.zip](https://github.com/visose/Robots/files/7927844/pp2.zip)

**Comment by visose** [comment, marked as answer]
<img width="475" alt="image" src="https://user-images.githubusercontent.com/5129614/150880400-f13bce12-79d1-4695-83d8-fca968ab5508.png">

This means the same variable name appears twice. This is likely because you have two .dat files in your robot in different places. Delete all the previous files generated with Grasshopper from the robot.

**Reply by alexmkrolak** [reply]
Thanks visose, this fixed it. Can you think of a way to avoid having to delete previous files on the robot?

**Reply by visose** [reply]
What is your particular use case for needing to store multiple programs generated in Grasshopper in the robot controller at the same time? Serious question, as I guess there could be a need for this, but in my past experience, these other programs have always been older versions of the same program that will never be run again. Or of someone else sharing the robot that will be loading a new program anyway when he gets to use the robot again.

There could be cases where you might create multiple programs that you don't want to run immediately in a sequence. Like one program to help setup your environment or calibrate, and a second one being the actual program, but is still best to copy the first one, once everything is working fine, copy the next one.

If I remember correctly, the reasoning behind using global variables in the .dat file was so that we could create multiple subprograms (multiple .src files that are run in a sequence) and they would all be able to read the variables from a single .dat file. The alternative was to create one .dat file per .src file, but we wanted things like speed, frame, tool variables defined only once in a single place.

There might be a better way to not expose global variables that could interfere with other programs stored in the controller, but I found the unintended consequence of forcing disorganized people to clean up unnecessary files from the controller a pleasant surprise.
